### ISSUE TITLE
Unable to update Bond due to CHANGE_BOND action: Invalid resulting order "3.00000023842".

### ISSUE THREAD

**Issue by mjohnson541** [ReactionMechanismGenerator/RMG-Py#1157]
I think this might be a rounding error?

```
Traceback (most recent call last):
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmg.py", line 162, in <module>
    rmg.execute(**kwargs)
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/rmg/main.py", line 721, in execute
    bimolecularReact=self.bimolecularReact)
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/rmg/model.py", line 658, in enlarge
    self.processNewReactions([rxn], spc)
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/rmg/model.py", line 725, in processNewReactions
    rxn, isNew = self.makeNewReaction(rxn)
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/rmg/model.py", line 476, in makeNewReaction
    products  = [self.makeNewSpecies(product)[0]  for product  in forward.products ]
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/rmg/model.py", line 326, in makeNewSpecies
    submit(spec)
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/thermo/thermoengine.py", line 179, in submit
    spc.thermo = submit_(evaluator, spc)
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/scoop_framework/util.py", line 171, in submit_
    return func(*args, **kwargs)
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/thermo/thermoengine.py", line 165, in evaluator
    thermo = generateThermoData(spc)
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/thermo/thermoengine.py", line 129, in generateThermoData
    thermo0 = thermodb.getThermoData(spc) 
  File "/Users/mattjohnson/RMGCODE/RMG-Py/rmgpy/data/thermo.py", line 1140, in getThermoData
    thermo0.S298.value_si -= constants.R * math.log(species.getSymmetryNumber())
  File "rmgpy/species.py", line 401, in rmgpy.species.Species.getSymmetryNumber
  File "rmgpy/species.py", line 467, in rmgpy.species.Species.getResonanceHybrid
  File "rmgpy/molecule/molecule.py", line 650, in rmgpy.molecule.molecule.Bond.applyAction
  File "rmgpy/molecule/molecule.py", line 636, in rmgpy.molecule.molecule.Bond.__changeBond
rmgpy.exceptions.ActionError: Unable to update Bond due to CHANGE_BOND action: Invalid resulting order "3.00000023842".
```

**Comment by goldmanm** [comment]
A fix has been implemented on #1113

**Comment by mjohnson541** [comment]
Is 6654a1e safe to cherry-pick?

**Comment by rwest** [comment]
OK, so my fears expressed in the comment thread on https://github.com/ReactionMechanismGenerator/RMG-Py/pull/864#issuecomment-276191600 have come home to roost. Can anyone think of anything we can do to protect against this type of bug happening again?

**Comment by mjohnson541** [comment]
We could create a BondOrder(float) class that has a looser definition of equality, but that adds an additional layer of complexity that may be undesirable.

**Comment by mjohnson541** [comment]
That would look something like:  
class BondOrder(float):
    def __init__(self,value):
        float.__init__(value)
    def __eq__(self,other):
        return abs(self-other)<.0001
    def __add__(self,other):
        return BondOrder(self.real+other.real)
    def __mul__(self,other):
        return BondOrder(self.real*other.real)

**Comment by goldmanm** [comment]
Fortunately, the error occurred within the Bond class, so once it is fixed, every location calling it will also be fixed. 

I am not sure of a fool-proof method. Adding an extra class layer could also cause complexity issues, especially when we are comparing bonds and GroupBonds (though your inheritance seems to reduce them). For example, comparing BondOrder with float, int, and other classes may be difficult to get correct. There are already similar methods for comparing in the Bond class (`equivalent`, `isOrder`,...)

One change we could make is to make the bond attribute private, which would require users to use the built-in comparison methods in the Bond class. There is currently an equivalent of isEqual in the Bond class. Do you think it might be beneficial to add isHigher and isLower, which can be used both inside and outside the class to better compare inequality?

**Comment by mjohnson541** [comment]
I don't know about GroupBonds, but the above inheritance structure appears to compare with int and floats as desired automatically.

**Comment by rwest** [comment]
A benefit of the builtin float/int/sting types is that we can do very fast comparisons and operations in the C layer of cythonized code. A special class may cause a significant performance hit, as well as adding some complexity. Cythonizing the class may regain some of the speed (at a cost of even more complexity).
Hmm. Maybe there is no elegant universal solution except remain vigilant and have good unit tests.

**Comment by goldmanm** [comment]
this will be fixed whenever #1171 gets merged

**Comment by mliu49** [comment]
Closed by #1171.
